test/net/sourceforge/kolmafia/session/RelayAutomationTest.java

```diff
@@ -0,0 +1,245 @@
+package net.sourceforge.kolmafia.session;
+
+import static internal.helpers.Networking.assertGetRequest;
+import static internal.helpers.Networking.assertPostRequest;
+import static internal.helpers.Networking.html;
+import static internal.helpers.Player.withHttpClientBuilder;
+import static internal.helpers.Player.withPasswordHash;
+import static internal.helpers.Player.withPath;
+import static internal.helpers.Player.withQuestProgress;
+import static internal.matchers.Quest.isStep;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.greaterThan;
+import static org.hamcrest.Matchers.hasSize;
+
+import internal.helpers.Cleanups;
+import internal.network.FakeHttpClientBuilder;
+import java.util.List;
+import java.util.Map;
+import net.sourceforge.kolmafia.AscensionPath;
+import net.sourceforge.kolmafia.AscensionPath.Path;
+import net.sourceforge.kolmafia.KoLAdventure;
+import net.sourceforge.kolmafia.KoLCharacter;
+import net.sourceforge.kolmafia.persistence.AdventureDatabase;
+import net.sourceforge.kolmafia.persistence.QuestDatabase.Quest;
+import net.sourceforge.kolmafia.preferences.Preferences;
+import net.sourceforge.kolmafia.request.GenericRequest;
+import net.sourceforge.kolmafia.request.RelayRequest;
+import net.sourceforge.kolmafia.webui.RelayAgent;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ValueSource;
+
+public class RelayAutomationTest {
+
+  @BeforeAll
+  public static void beforeAll() {
+    // Simulate logging out and back in again.
+    GenericRequest.passwordHash = "";
+    KoLCharacter.reset("");
+    KoLCharacter.reset("relay automation");
+    Preferences.saveSettingsToFile = false;
```

Comment:
This is handled in `ClearSharedStateBefore` and shouldn't be necessary in each test.

Reply:
To be more specific, the `saveSettingsToFile` change is handled in that class (among other things I haven't checked).

Reply:
Huh. I have not looked at that before. I see that it does the following:
```
   GenericRequest.passwordHash = "";
   KoLCharacter.reset("");
   Preferences.saveSettingsToFile = false;
 ```
I notice that the ```Preferences.saveSettingToFIle``` is NOT restored to true (the default) in the "After" extension.

I think that
```
    KoLCharacter.reset("relay automation");
    Preferences.reset("relay automation");
```
Need to be done by individual test files in BeforeAll and BeforeEach respectively.

Now that I have studied (a bit) the extension, I'll remove the 3 unnecessary lines and will pay attention in the future.
Although, there are plenty of test files that could be cleaned up. And also be djusted to use CleanUps. :)

Reply:
It is done.